lib/codecs/src/decoding/framing/chunked_gelf.rs

```diff
@@ -0,0 +1,692 @@
+    InvalidTotalChunks {
+        message_id: u64,
+        sequence_number: u8,
+        total_chunks: u8,
+    },
+    #[snafu(display("Sequence number {sequence_number} is greater than the total chunks value {total_chunks} for message with id {message_id}"))]
+    InvalidSequenceNumber {
+        message_id: u64,
+        sequence_number: u8,
+        total_chunks: u8,
+    },
+    #[snafu(display("Pending messages limit of {pending_messages_limit} reached while processing message with id {message_id} and sequence number {sequence_number}"))]
+    PendingMessagesLimitReached {
+        message_id: u64,
+        sequence_number: u8,
+        pending_messages_limit: usize,
+    },
+    #[snafu(display("Received message with id {message_id} and sequence number {sequence_number} has different total chunks values: original total chunks value {original_total_chunks}, received total chunks value {received_total_chunks}"))]
+    TotalChunksMismatch {
+        message_id: u64,
+        sequence_number: u8,
+        original_total_chunks: u8,
+        received_total_chunks: u8,
+    },
+}
+
+impl StreamDecodingError for ChunkedGelfDecoderError {
+    fn can_continue(&self) -> bool {
+        true
+    }
+}
+
+impl FramingError for ChunkedGelfDecoderError {
+    fn as_any(&self) -> &dyn Any {
+        self as &dyn Any
+    }
+}
+
+/// A codec for handling GELF messages that may be chunked. The implementation is based on [Graylog's GELF documentation](https://go2docs.graylog.org/5-0/getting_in_log_data/gelf.html#GELFviaUDP)
+/// and [Graylog's go-gelf library](https://github.com/Graylog2/go-gelf/blob/v1/gelf/reader.go).
+#[derive(Debug, Clone)]
+pub struct ChunkedGelfDecoder {
+    // We have to use this decoder to read all the bytes from the buffer first and don't let tokio
+    // read it buffered, as tokio FramedRead will not always call the decode method with the
+    // whole message. (see https://docs.rs/tokio-util/latest/src/tokio_util/codec/framed_impl.rs.html#26).
+    // This limitation is due to the fact that the GELF format does not specify the length of the
+    // message, so we have to read all the bytes from the message (datagram)
+    bytes_decoder: BytesDecoder,
+    state: Arc<Mutex<HashMap<u64, MessageState>>>,
+    timeout: Duration,
+    pending_messages_limit: usize,
+}
+
+impl ChunkedGelfDecoder {
+    /// Creates a new `ChunkedGelfDecoder`.
+    pub fn new(timeout_secs: f64, pending_messages_limit: usize) -> Self {
+        Self {
+            bytes_decoder: BytesDecoder::new(),
+            state: Arc::new(Mutex::new(HashMap::new())),
+            timeout: Duration::from_secs_f64(timeout_secs),
+            pending_messages_limit,
+        }
+    }
+
+    /// Decode a GELF chunk
+    pub fn decode_chunk(
+        &mut self,
+        mut chunk: Bytes,
+    ) -> Result<Option<Bytes>, ChunkedGelfDecoderError> {
+        // Encoding scheme:
```

Comment:
I think it would be useful to record a metric for tracking the number of current pending messages. How can we approach this? I think this is usually done at Vector's binary level and not in inner libs... as I see a lot of metrics located at `src/internal_events`.

Also, it would be useful to record the number of timed out messages

Reply:
That would be useful telemetry. Unfortunately I don't think there is a precedent for codecs emitting metrics, so this may require a bit of scaffolding if you want to add it. I don't think we want to emit telemetry directly from the codec crate so it'd look something like exposing metrics that then the caller (Vector) can emit.

I think this could mean adding a trait method to `Framer`, https://github.com/vectordotdev/vector/blob/17466c684bc6d3b0356c976e9639f9b05d306769/lib/codecs/src/decoding/framing/mod.rs#L67-L81, that exposes telemetry that the caller than polls and emits; or maybe it takes a callback that is called when the codec wants to emit telemetry (example from the [kafka source](https://github.com/vectordotdev/vector/blob/17466c684bc6d3b0356c976e9639f9b05d306769/src/sources/kafka.rs#L1365-L1369))

Reply:
And, yes, we generally try to have only `internal_events` actually emit metrics to centralize it all in one place.

Reply:
I think emitting metrics from this point would require a refactor that in my opinion is slightly off from this PR. Maybe we can create an issue to track metrics in decoders in general.

I don't mind to address it myself as I found it interesting, but I prefer to narrow the scope of this PR